kimchi/src/circuits/polynomials/range_check/witness.rs

```diff
@@ -79,152 +24,118 @@ impl ZeroWitnessCell {
-        LimbWitnessCell::create(2, 0, 2, 4),
-        LimbWitnessCell::create(2, 0, 0, 2),
-    ],
-];
+        /* row 1, RangeCheck0 row */
+        range_check_0_row("v0", 0),
+        /* row 2, RangeCheck0 row */
+        range_check_0_row("v1", 1),
+        /* row 3, RangeCheck1 row */
+        [
+            VariableCell::create("v2"),
+            /* 2-bit crumbs (placed here to keep lookup pattern */
+            /*               the same as RangeCheck0) */
+            CopyBitsCell::create(2, 0, 86, 88),
+            CopyBitsCell::create(2, 0, 84, 86),
+            /* 12-bit plookups */
+            CopyBitsCell::create(2, 0, 72, 84),
+            CopyBitsCell::create(2, 0, 60, 72),
+            CopyBitsCell::create(2, 0, 48, 60),
+            CopyBitsCell::create(2, 0, 36, 48),
+            /* 2-bit crumbs */
+            CopyBitsCell::create(2, 0, 34, 36),
+            CopyBitsCell::create(2, 0, 32, 34),
+            CopyBitsCell::create(2, 0, 30, 32),
+            CopyBitsCell::create(2, 0, 28, 30),
+            CopyBitsCell::create(2, 0, 26, 28),
+            CopyBitsCell::create(2, 0, 24, 26),
+            CopyBitsCell::create(2, 0, 22, 24),
+            CopyBitsCell::create(2, 0, 20, 22),
+        ],
+        /* row 4, Zero row */
+        [
+            ConstantCell::create(F::zero()),
+            /* 2-bit crumbs (placed here to keep lookup pattern */
+            /*               the same as RangeCheck0) */
+            CopyBitsCell::create(2, 0, 18, 20),
+            CopyBitsCell::create(2, 0, 16, 18),
+            /* 12-bit plookups (see note about copies in range_check_row) */
+            CopyCell::create(0, 1),
+            CopyCell::create(0, 2),
+            CopyCell::create(1, 1),
+            CopyCell::create(1, 2),
+            /* 2-bit crumbs */
+            CopyBitsCell::create(2, 0, 14, 16),
+            CopyBitsCell::create(2, 0, 12, 14),
+            CopyBitsCell::create(2, 0, 10, 12),
+            CopyBitsCell::create(2, 0, 8, 10),
+            CopyBitsCell::create(2, 0, 6, 8),
+            CopyBitsCell::create(2, 0, 4, 6),
+            CopyBitsCell::create(2, 0, 2, 4),
+            CopyBitsCell::create(2, 0, 0, 2),
+        ],
+    ]
+}
 
 /// The row layout for `RangeCheck0`
-const fn range_check_row(row: usize) -> [WitnessCell; COLUMNS] {
+fn range_check_0_row<F: PrimeField>(
+    limb_name: &'static str,
+    row: usize,
+) -> [Box<dyn WitnessCell<F>>; COLUMNS] {
     [
-        ValueWitnessCell::create(),
+        VariableCell::create(limb_name),
         /* 12-bit copies */
         // Copy cells are required because we have a limit
         // of 4 lookups per row.  These two lookups are moved to
         // the 4th row, which is a Zero circuit gate, and the
         // RangeCheck1 circuit gate triggers the lookup constraints.
-        LimbWitnessCell::create(row, 0, 76, 88),
-        LimbWitnessCell::create(row, 0, 64, 76),
+        CopyBitsCell::create(row, 0, 76, 88),
```

Comment:
This would be a bit more readable (for me) as `Box::create(CopyBitsCell {row, col: 0, start:76, end:88})`. Likewise elsewhere.